source/code/source/playfab/PlayFabEventPipeline.cpp

```diff
@@ -166,8 +166,15 @@ namespace PlayFab
         {
             try
             {
+                size_t sizeOfBatchesInFlight = 0;
+
+                { // LOCK batchesInFlight mutex
+                    std::unique_lock<std::mutex> lock(inFlightMutex);
+                    sizeOfBatchesInFlight = batchesInFlight.size();
+                } // UNLOCK batchesInFlight
+
                 // Process events in the loop
-                if (this->batchesInFlight.size() >= this->settings->maximalNumberOfBatchesInFlight)
+                if (sizeOfBatchesInFlight >= this->settings->maximalNumberOfBatchesInFlight)
```

Comment:
do all of the references to `this->settings` need to be reevaluated for thread safety in this file?
It looks like the event pipeline exposes a public `GetSettings` method which returns a mutable shared_ptr to the settings field. That enables external callers to modify the settings used by this event pipeline across any number of threads while the worker thread is running.

This can cause tearing of the state in the `settings` member which could lead to some unexpected behavior if anyone tries to mess with it while the worker thread is running.

unfortunately the `this->settings` value is used to keep track of the auth context needed in the event pipeline so I imagine making it thread safe could prove difficult... (but not impossible!)

Reply:
This change is starting to get big. I have created a different bug to track this particular issue: https://dev.azure.com/playfab/PlayFab/_workitems/edit/42685

Reply:
that seems super reasonable to me! modifying the settings in the middle of the event api running is probably not an intended calling pattern and it's not a regression so I wouldn't prioritize it for this change.

but I think it's worth tracking as a potential place customers could misuse the api. thanks for filing the bug. I'll resolve this thread.